algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py

```diff
@@ -0,0 +1,253 @@
+    result[result == -1] = 0
+    result = result[:, :-1]
+
+    # remove blanks (id = 0)
+    blank_id = 0
+    fin_result = torch.zeros_like(result)
+    idxs = torch.arange(
+        fin_result.numel(), device=result.device).view(*fin_result.shape)
+    mask = torch.arange(
+        fin_result.shape[1], device=result.device).view(
+            1, -1) < result.count_nonzero(dim=1).view(-1, 1)
+    fin_result.view(-1)[idxs[mask != 0]] = result[result != blank_id]
+    padding = (fin_result == 0)
+    return fin_result, padding
+
+  def _eval_model_on_split(self,
+                           split: str,
+                           num_examples: int,
+                           global_batch_size: int,
+                           params: spec.ParameterContainer,
+                           model_state: spec.ModelAuxiliaryState,
+                           rng: spec.RandomState,
+                           data_dir: str,
+                           global_step: int):
+    """Run a full evaluation of the model."""
+    data_rng, model_rng = prng.split(rng, 2)
+    if split not in self._eval_iters:
+      # These iterators repeat indefinitely.
+      self._eval_iters[split] = itertools.cycle(
+          self.build_input_queue(
+              data_rng, split, data_dir, global_batch_size=global_batch_size))
+
+    total_metrics = {
+        'ctc_loss': torch.tensor(0., device=DEVICE),
+        "word_errors": torch.tensor(0., device=DEVICE),
+        "num_words": torch.tensor(0., device=DEVICE),
+    }
+    num_batches = int(math.ceil(num_examples / global_batch_size))
+
+    for _ in range(num_batches):
+      batch = next(self._eval_iters[split])
+
+      (logits, logits_padding), _ = self.model_fn(
+          params,
+          batch,
+          model_state,
+          spec.ForwardPassMode.EVAL,
+          model_rng,
+          update_batch_norm=False)
+      decoded, decoded_paddings = self.greedy_decode(logits, logits_padding)
+      word_errors, num_words = metrics.compute_wer(
+          decoded=decoded.detach().cpu().numpy(),
+          decoded_paddings=decoded_paddings.detach().cpu().numpy(),
+          targets=batch['targets'].detach().cpu().numpy(),
+          target_paddings=batch["target_paddings"].detach(
+          ).cpu().numpy(),
+          tokenizer=self.tokenizer)
+      batch_metrics = {
+          "ctc_loss":
+              self.compute_loss(logits,
+                                logits_padding,
+                                batch['targets'],
+                                batch["target_paddings"]).sum(),
+          "word_errors":
+              word_errors,
+          "num_words":
+              num_words
+      }
+      total_metrics = {
+          k: v + batch_metrics[k] for k, v in total_metrics.items()
+      }
+    if USE_PYTORCH_DDP:
+      for metric in total_metrics.values():
+        dist.all_reduce(metric)
+    return {
+        "ctc_loss": float(total_metrics["ctc_loss"].item() / num_examples),
```

Comment:
if my above comment about numerator / denominator is mistaken, and we keep 'ctc_loss' as a key, I think this needs to be ` / num_batches` not num_examples?

Reply:
Good catch! But, I think the denominator has to be the number of gpus because each gpu has the average loss for the batch slice it received? I'll clean all this up soon!

Reply:
I cleaned this up by returning a dictionary from the compute_loss function: i have only implemented a dummy `loss_fn` as of now.